Zend/zend_language_parser.y

```diff
@@ -422,21 +429,55 @@ is_variadic:
 ;
 
 class_declaration_statement:
-		class_type { $<num>$ = CG(zend_lineno); }
+		class_modifiers T_CLASS { $<num>$ = CG(zend_lineno); }
 		T_STRING extends_from implements_list backup_doc_comment '{' class_statement_list '}'
-			{ $$ = zend_ast_create_decl(ZEND_AST_CLASS, $1, $<num>2, $6,
-				  zend_ast_get_str($3), $4, $5, $8); }
-	|	T_INTERFACE { $<num>$ = CG(zend_lineno); }
-		T_STRING interface_extends_list backup_doc_comment '{' class_statement_list '}'
-			{ $$ = zend_ast_create_decl(ZEND_AST_CLASS, ZEND_ACC_INTERFACE, $<num>2, $5,
-				  zend_ast_get_str($3), NULL, $4, $7); }
+			{
+				if (!($1 & ZEND_ACC_PPP_MASK)) { $1 |= ZEND_ACC_PUBLIC; }
+				$$ = zend_ast_create_decl(ZEND_AST_CLASS, $1, $<num>3, $7, zend_ast_get_str($4), $5, $6, $9); 
+			}
+	|	T_CLASS { $<num>$ = CG(zend_lineno); }
+		T_STRING extends_from implements_list backup_doc_comment '{' class_statement_list '}'
+			{ $$ = zend_ast_create_decl(ZEND_AST_CLASS, ZEND_ACC_PUBLIC, $<num>2, $6, zend_ast_get_str($3), $4, $5, $8); }
+;
+
+class_modifiers:
+		class_modifier 					{ $$ = $1; }
+	|	class_modifiers class_modifier 	{ $$ = zend_add_class_modifier($1, $2); }
+;
+
+class_modifier:
+		T_ABSTRACT 				{ $$ = ZEND_ACC_EXPLICIT_ABSTRACT_CLASS; }
+	|	T_FINAL 				{ $$ = ZEND_ACC_FINAL; }
+	|	visibility_modifier 	{ $$ = $1; }
+;
+
+visibility_modifier:
+		T_PUBLIC 		{ $$ = ZEND_ACC_PUBLIC; }
+	|	T_PRIVATE 		{ $$ = ZEND_ACC_PRIVATE; }
```

Comment:
I think, let's not waste energy on this now, and we can state very clearly
that this can be a future RFC
On Feb 19, 2015 5:24 PM, "Guilherme Blanco" <notifications@github.com>
wrote:

> In Zend/zend_language_parser.y
> <https://github.com/php/php-src/pull/947#discussion_r25006217>:
>
> > +;
> > +
> > +class_modifiers:
> > +		class_modifier 					{ $$ = $1; }
> > +	|	class_modifiers class_modifier 	{ $$ = zend_add_class_modifier($1, $2); }
> > +;
> > +
> > +class_modifier:
> > +		T_ABSTRACT 				{ $$ = ZEND_ACC_EXPLICIT_ABSTRACT_CLASS; }
> > +	|	T_FINAL 				{ $$ = ZEND_ACC_FINAL; }
> > +	|	visibility_modifier 	{ $$ = $1; }
> > +;
> > +
> > +visibility_modifier:
> > +		T_PUBLIC 		{ $$ = ZEND_ACC_PUBLIC; }
> > +	|	T_PRIVATE 		{ $$ = ZEND_ACC_PRIVATE; }
>
> @lt <https://github.com/lt> personally I don't think this is valid.
> Let's check an example:
>
> protected class Foo {}namespace A {    class B extends A {}}
>
> This example should be valid too? Since you allow deep namespacing usage,
> that would be valid.
> Now, a namespace should only have access to either its siblings or its
> internal namespaces, but shouldn't rely on parent namespaces. Taking this
> approach promotes breaking encapsulation and this is very bad. That's why
> C# and Java does not support any other kind of class visibility at top
> level.
>
> Now there's one thing I don't know if it should work. Suppose I have a
> base Factory that instantiates concrete implementations of an interface.
> Interface is public, so it would allow external implementors to add new
> behavior. Now if I add a new support, I would be required to extend the
> Factory to add my newly created concrete implementation instantiation.
> Since I'd be calling parent::create(...) to instantiate the other
> concretes, should it allow to be created (if they are private classes) or
> should it deny?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/php/php-src/pull/947/files#r25006217>.
>